File: src/test/java/composing/methods/extractMethods.java
package composing.methods;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class extractMethods {
    private final ArrayList<Product> products = new ArrayList<>(
            List.of(Product.builder().name("Blu Ray Tenet").price(31.0).build(),
                    Product.builder().name("Book Fundamentals of Software Architecture").price(54.22).build(),
                    Product.builder().name("T-Shirt Geek Kaamelott").price(25.90).build())
    );
    private final Customer customer = Customer.builder().age(35).name("Mando").build();
    private final Order order = Order.builder().products(products).customer(customer).build();

    @Test
    public void generateStatement() {
        assertEquals("Statement for : Customer{name='Mando', age=35}\n" +
                "Product: Blu Ray Tenet Price: 31.0\n" +
                "Product: Book Fundamentals of Software Architecture Price: 54.22\n" +
                "Product: T-Shirt Geek Kaamelott Price: 25.9\n" +
                "Total: 111.12€", order.generateStatement());
    }

    @Test
    public void calculatePrice() {
        assertEquals(9.425, AmountCalculator.calculatePrice(14.5, 16));
        assertEquals(9.425, AmountCalculator.calculatePrice(14.5, 12));
        assertEquals(14.5, AmountCalculator.calculatePrice(14.5, 30));
        assertEquals(11.6, AmountCalculator.calculatePrice(14.5, 60));
        assertEquals(11.6, AmountCalculator.calculatePrice(14.5, 80));
    }

    @Test
    public void calculatePriceForOrders() {
        assertEquals(72.22, AmountCalculator.calculatePrice(order, true, 16), 0.01);
        assertEquals(111.12, AmountCalculator.calculatePrice(order, false, 16), 0.01);
        assertEquals(111.12, AmountCalculator.calculatePrice(order, true, 40), 0.01);
        assertEquals(88.89, AmountCalculator.calculatePrice(order, true, 70), 0.01);
        assertEquals(111.12, AmountCalculator.calculatePrice(order, false, 70), 0.01);
    }
}

File: src/scala/src/test/scala/composing/methods/extractMethods.scala
package composing.methods

import composing.methods.AmountCalculator.calculatePrice
import org.scalactic.Tolerance.convertNumericToPlusOrMinusWrapper
import org.scalatest.funsuite.AnyFunSuite

import scala.util.Success

class extractMethods extends AnyFunSuite {
  private val products: List[Product] =
    List(
      Product("Blu Ray Tenet", 31.0),
      Product("Book Fundamentals of Software Architecture", 54.22),
      Product("T-Shirt Geek Kaamelott", 25.90)
    )

  private val customer: Customer = Customer("Mando", 35)
  private val order: Order = Order(customer, products)

  test("generate statement") {
    assert(
      order.generateStatement == Success(
        """Statement for : Customer{name='Mando', age=35}
        |Product: Blu Ray Tenet Price: 31.0
        |Product: Book Fundamentals of Software Architecture Price: 54.22
        |Product: T-Shirt Geek Kaamelott Price: 25.9
        |Total: 111.12€""".stripMargin
      )
    )
  }

  test("calculate price") {
    assertPrice(14.5, 16, 9.425)
    assertPrice(14.5, 12, 9.425)
    assertPrice(14.5, 30, 14.5)
    assertPrice(14.5, 60, 11.6)
    assertPrice(14.5, 80, 11.6)
  }

  private def assertPrice(amount: Double, age: Int, expectedResult: Double) = {
    assert(calculatePrice(amount, age) == expectedResult)
  }

  test("calculatePriceForOrders") {
    assertOrderPrice(applyAgeDiscount = true, 16, 72.23)
    assertOrderPrice(applyAgeDiscount = false, 16, 111.12)
    assertOrderPrice(applyAgeDiscount = true, 40, 111.12)
    assertOrderPrice(applyAgeDiscount = true, 70, 88.90)
    assertOrderPrice(applyAgeDiscount = false, 70, 111.12)
  }

  private def assertOrderPrice(
      applyAgeDiscount: Boolean,
      age: Int,
      expectedResult: Double
  ) = {
    assert(
      calculatePrice(order, applyAgeDiscount, age) === expectedResult +- 0.01
    )
  }
}
